src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcApplicationBuilderExtensions.cs

```diff
@@ -89,7 +89,7 @@ public static IApplicationBuilder UseMvcWithDefaultRoute(this IApplicationBuilde
 
             var routes = new RouteBuilder(app)
             {
-                DefaultHandler = new MvcRouteHandler(),
+                DefaultHandler = app.ApplicationServices.GetRequiredService<IRouter>(),
```

Comment:
Register this directly as `MvcRouteHandler` - it's not the be-all-end-all of routing